src/harddisk/fat32.rs

```diff
@@ -0,0 +1,1554 @@
+		}
+
+		let sectors_per_cluster = sector[0x0D] as _;
+
+		let reserved_sectors = u16::from_le_bytes([sector[0x0E], sector[0x0F]]) as _;
+
+		let fat_count = sector[0x10] as _;
+
+		let root_dir_entries = u16::from_le_bytes([sector[0x11], sector[0x12]]) as _;
+
+		let total_sectors = u16::from_le_bytes([sector[0x13], sector[0x14]]) as _;
+
+		let sectors_per_fat = if let FatVersion::Fat32 { .. } = version {
+			u32::from_le_bytes([sector[0x24], sector[0x25], sector[0x26], sector[0x27]])
+		} else {
+			let count = u16::from_le_bytes([sector[0x16], sector[0x17]]) as u32;
+			if count == 0 {
+				u32::from_le_bytes([sector[0x20], sector[0x21], sector[0x22], sector[0x23]])
+			} else {
+				count
+			}
+		} as _;
+
+		let label = {
+			let addr = match version {
+				FatVersion::Fat32 { .. } => 0x47,
+				_ => 0x2B,
+			};
+
+			let mut label = SVec::new();
+			for b in &sector[addr..addr + 11] {
+				label.push(*b);
+			}
+
+			while label[label.len() - 1] == b' ' {
+				label.pop();
+			}
+
+			label
+		};
+
+		Ok(Header {
+			oem_ident,
+			sectors_per_cluster,
+			reserved_sectors,
+			fat_count,
+			root_dir_entries,
+			total_sectors,
+			sectors_per_fat,
+			label,
+			fat_version: version,
+		})
+	}
+}
+
+struct Driver {
+	partition: usize,
+	header: Header,
+	fat: FileAllocationTable,
+	/// FAT sector
+	current_loaded_sector: usize,
+	/// Unlike the `fat`, there is no worry of breaching sector boundries here
+	buffer: [u8; 512],
+}
+
+static mut DRIVER: Driver = Driver::uninititalized();
+
+impl Driver {
+	const fn uninititalized() -> Self {
+		Self {
+			partition: 0,
+			header: Header {
+				oem_ident: SVec::new(),
+				sectors_per_cluster: 0,
+				reserved_sectors: 0,
+				fat_count: 0,
+				sectors_per_fat: 0,
+				label: SVec::new(),
+				root_dir_entries: 0,
+				fat_version: FatVersion::Fat12,
+				total_sectors: 0,
+			},
+			fat: FileAllocationTable {
+				version: FatVersion::Fat12,
+				sector_count: 0,
+				fat_offset: 0,
+				currently_loaded_sector: 0,
+				buffer: [0; 1024],
+			},
+			current_loaded_sector: 0,
+			buffer: [0; 512],
+		}
+	}
+
+	/// Initilize the driver
+	///
+	/// # Safety
+	///
+	/// There is no check if the driver has already been initialized
+	///
+	/// Requires partitions to already be initialized.
+	unsafe fn initialize(&mut self) {
+		for part in super::partitions::list_partitions() {
+			let start = part.start_sector();
+			let mut sector = [0; 512];
+			super::partitions::read_sectors(part.index(), 0, &mut sector);
+			if let Ok(header) = Header::try_new(&sector) {
+				self.header = header;
+				self.fat = FileAllocationTable::new(
+					self.header.fat_version,
+					self.header.total_sectors,
+					self.header.reserved_sectors,
+				);
+				super::partitions::read_sectors(part.index(), 0, &mut self.buffer);
+				// println!("{:#?}", self.header);
+				break;
+			}
+			// todo!("Check if sector counts are the same for the header and the partition");
```

Comment:
I believe this is done in header init, row 267, which means this should be removed